apps/omg_api/lib/monitor.ex

```diff
@@ -0,0 +1,152 @@
+# Copyright 2018 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.API.Monitor do
+  @moduledoc """
+  This module is a custom implemented supervisor that monitors all it's chilldren
+  and restarts them based on alarms raised. This means that in the period when Geth alarms are raised
+  it would wait before it would restart them.
+
+  When you receive an EXIT, check for an alarm raised that's related to Ethereum client synhronisation or connection
+  problems and react accordingly.
+
+  Children that need Ethereum client connectivity are OMG.API.EthereumEventListener
+  OMG.API.BlockQueue.Server and OMG.API.RootChainCoordinator. For these children, we make
+  additional checks if they exit. If there's an alarm raised of type :ethereum_client_connection we postpone
+  the restart util the alarm is cleared. Other children are restarted immediately.
+
+  """
+  use GenServer
+  require Logger
+  alias OMG.API.Alert.Alarm
+  # needs to be less then checks from RootChainCoordinator
+  @default_interval 300
+  @type t :: %__MODULE__{
+          pid: pid(),
+          spec: {module(), term()} | map(),
+          tref: :timer.tref() | nil
+        }
+  defstruct pid: nil, spec: nil, tref: nil
+
+  def start_link(children_specs) do
+    GenServer.start_link(__MODULE__, children_specs, name: __MODULE__)
+  end
+
+  def init(children_specs) do
+    Process.flag(:trap_exit, true)
+
+    children =
+      Enum.map(
+        children_specs,
+        fn spec ->
+          start_child(spec)
+        end
+      )
+
+    {:ok, children}
+  end
+
+  def handle_info({:delayed_restart, child}, state) do
+    # child still holds the old pid
+    case is_raised?() do
+      true ->
+        {:noreply, state}
+
+      _ ->
+        # old pid that the crash came from
+        from = child.pid
+        {%__MODULE__{pid: ^from, tref: tref} = child, other_children} = find_child_from_dead_pid(from, state)
```

Comment:
Not sure I understand. Restarts per linked process are synchronous.
1. we receive an exit from child **A** that needs `:ethereum_client_connection`
2. we check if there's an alarm raised from the ethereum client monitor
3. yes it is - we delay the restart for that child **A** and setup a timer
4. the timer kicks in and we retrieve the child **A** from the state with a key being the dead PID. At the time when the child died, we also set the timer ref back to the state.

Reply:
`{:delayed_restart, child}` is emitted by `:timer.send_interval`. Imagine message queue of API.Monitor having 2 such messages. First will cleanup `child` from `state`. Second one will crash since `find_child_from_dead_pid` will return `nil`.